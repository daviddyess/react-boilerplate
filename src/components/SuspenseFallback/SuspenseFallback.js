import React, { Component } from 'react';
import { Spinner } from 'react-bootstrap';
import Header from 'components/Header';
import Footer from 'components/Footer';

export default class SuspenseFallback extends Component {
  render() {
    return (
      <div>
        <Header />
        <div className="text-center p-4">
          <div className="m-5">
            <Spinner animation="border" role="status" />
            <div>
              <span>Loading...</span>
            </div>
          </div>
        </div>
        <Footer />
      </div>
    );
  }
}
