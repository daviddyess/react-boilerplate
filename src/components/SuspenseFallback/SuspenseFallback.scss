@import "../../variables";
$spinner-size: 15rem;

.loading {
  &__overlay {
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    background: rgba($lighter-blue, 0.75);
    z-index: 999;
  }

  &__content {
    position: relative;
    width: $spinner-size;
    height: $spinner-size;
    transform: translateY(40vh);
  }

  &__text {
    position: fixed;
    font-size: 2rem;
    width: $spinner-size;
    height: $spinner-size;
    text-align: center;

    span {
      display: inline-block;
      transform: translateY(6rem);
    }
  }

  &__spinner {
    position: fixed;
    border-color: $teal;
    border-width: 0.3em;
    border-right-color: transparent;
    width: $spinner-size;
    height: $spinner-size;
  }
}
