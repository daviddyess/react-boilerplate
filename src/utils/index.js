/**
 * Creates an object containing action constants namespaced under the specified reducer.
 *
 * @param {string} reducer The name of the reducer or module
 * @param {string[]} actions A list of action names
 * @return {object} Object with action name keys and full action string values
 * @author ayan4m1 <https://github.com/ayan4m1>
 */
export function buildActions(reducer, actions) {
  const result = {};

  for (const action of actions) {
    result[action] = `${reducer}/${action}`;
  }

  return result;
}

/**
 * Interpolate strings from the `params` object into the `url`.
 *
 * @param {object} endpoint Object with `url` and optional `params` object
 * @author ayan4m1 <https://github.com/ayan4m1>
 */
export const buildUrl = endpoint => {
  // this is defined by webpack.DefinePlugin
  const baseUrl = process.env.REACT_APP_API_URL || 'localhost';
  const { url, params } = endpoint;

  if (!params) {
    return `${baseUrl}${url}`;
  }

  return Object.entries(params).reduce((resultUrl, entry) => {
    const [key, value] = entry;

    return resultUrl.replace(`{${key}}`, value);
  }, `${baseUrl}${url}`);
};