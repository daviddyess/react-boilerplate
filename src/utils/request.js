import axios from 'axios';
import { buildUrl } from '../utils';
import { call, race, delay } from 'redux-saga/effects';

/**
 * Based on code by:
 * @author ayan4m1 <https://github.com/ayan4m1>
 * As of 20191117
 * Modified by:
 * @author daviddyess <https://github.com/daviddyess>
 */

/**
 * Creates a response object indicating successful response from the server.
 *
 * @param {object} response The response body
 */
const successResponse = response => ({
  success: true,
  response
});

/**
 * Creates a response object indicating failure with one specific error message.
 *
 * @param {string} message A string indicating the error message.
 */
const failureMessage = message => ({
  success: false,
  error: {
    message
  }
});

/**
 * Creates a response object indicating failure with an Error or error object.
 *
 * @param {Error|object} error An object describing the error
 */
const failureResponse = error => ({
  success: false,
  error
});

/**
 * Make an XMLHttpRequest to the specified endpoint with the specified request data.
 *
 * @param {object} endpoint An object with `url` and `method` keys
 * @param {object} data An object to use as the request body
 */
class Request {
  constructor() {
    this.execute = this.execute.bind(this);
  }

  isUrlProtected(url) {
    const unprotectedResources = ['oauth', 'register'];

    for (const resource of unprotectedResources) {
      if (url.startsWith(`/${resource}`)) {
        return false;
      }
    }

    return true;
  }

  *execute({ endpoint, data, headers, options = {} }) {
    try {
      // ensure the endpoint was supplied
      if (!endpoint) {
        return failureMessage('No endpoint provided!');
      }

      // ensure properties required by axios are present
      const { url, method } = endpoint;

      if (!url) {
        return failureMessage('Endpoint is missing URL!');
      }

      if (!method) {
        return failureMessage('Endpoint is missing method!');
      }

      // default timeout of 10 seconds
      const { timeout = 10000 } = options;

      const requestUrl = yield call(buildUrl, endpoint);

      // start a race between the request and a timer, cancel the loser
      const { response } = yield race({
        response: call(axios, {
          url: requestUrl,
          headers,
          method,
          data
        }),
        timeout: delay(timeout)
      });

      if (!response) {
        return failureMessage('Request timed out!');
      }

      const { status } = response;
      const success = status >= 200 && status < 400;

      return success ? successResponse(response) : failureResponse(response);
    } catch (error) {
      return failureResponse(error);
    }
  }
}

const instance = new Request();

export default instance;
