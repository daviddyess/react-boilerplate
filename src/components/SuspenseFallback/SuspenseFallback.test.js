import React from 'react';
import renderer from 'react-test-renderer';

import SuspenseFallback from './SuspenseFallback';
import { withMemoryRouter } from 'utils';
jest.mock('components/Header/Header', () =>
  require('utils').mockComponent('Header')
);

describe('<SuspenseFallback />', () => {
  it('renders correctly', () => {
    const RoutedFallback = withMemoryRouter(SuspenseFallback);
    const component = renderer.create(<RoutedFallback />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
