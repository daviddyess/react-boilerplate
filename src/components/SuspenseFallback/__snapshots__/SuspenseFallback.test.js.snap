// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SuspenseFallback /> renders correctly 1`] = `
<div>
  <Header />
  <div
    className="loading__overlay"
  >
    <div
      className="loading__content m-auto"
    >
      <div
        className="loading__spinner spinner-border"
        role="status"
      />
      <div
        className="loading__text"
      >
        <span>
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>
`;
